### PR TITLE
Remove incremental

### LINKED ISSUE #308: Remove dependency on incremental in favor of general compatible solution
https://github.com/twisted/towncrier/pull/297#pullrequestreview-556014893
> For `get_version` I think that we should have this logic
> 
> See if version is a tuple and join them with a dot.
> For anything else, just call `str(version_object)` and strip it.
> 
> In this way it will work with incremental and any other version objects.
> 
> towncrier shouldn't care about comparing versions...
> 
> It can try to be smart and detect that you are not adding a version older than the latest one in the file... but I think that is too much...and again it should be implemented with a naive compare
> 
> So I am +1 for remove the incremental dependency :) and use simple version strings for towncrier :)
> 
> I am using a fork of towncrier in which incremental is removed as at least 3 years ago, incremental was not playing nice with private PYPI repos due to a bug in setup tools

Sounds like we could remove the dependency specifically on incremental.  Give it a shot if we get a chance.

### PR DESCRIPTION
**Pull request by SmileyChris**
# Description
Remove Incremental as a requirement (but still support its use for project versions/names)

Fixes #308 and #398

# Checklist

* [x] Make sure changes are covered by existing or new tests.
* [x] For at least one Python version, make sure local test run is green.
* [x] Create a file in `src/towncrier/newsfragments/`. Describe your
  change and include important information. Your change will be included in the public release notes.
* [x] Make sure all GitHub Actions checks are green (they are automatically checking all of the above).
* [x] Ensure `docs/tutorial.rst` is still up-to-date.
* [x] If you add new **CLI arguments** (or change the meaning of existing ones), make sure `docs/cli.rst` reflects those changes.
* [x] If you add new **configuration options** (or change the meaning of existing ones), make sure `docs/configuration.rst` reflects those changes.

### CODE REVIEW

File: src/towncrier/test/test_project.py [hunk 1]
```diff
@@ -48,25 +41,61 @@ def test_tuple(self):
...
-            DeprecationWarning, "Accessing towncrier.__version__ is deprecated.*"
-        ):
-            name = get_project_name(pkg, "towncrier")
 
-        self.assertEqual(metadata_version("towncrier"), version)
-        self.assertEqual("towncrier", name)
+__version__ = WeirdVersion()
+"""
+            )
+        with self.assertRaises(Exception) as e:
```
Comment (chrysle) [inline src/towncrier/test/test_project.py:90]: Which exception does this raise?

Comment (SmileyChris) [reply src/towncrier/test/test_project.py:90]: Literally [`Exception`](https://github.com/twisted/towncrier/pull/491/files#diff-ea234c597bb51dfad3d154ebdeb24af6c75e95ac9a922267f99e7704f1fbde42R72) -- the definition of that exception is existing code, but we can of course tighten this up if it makes sense now.

**Comment by adiroiban** [comment]
Thanks, Chris for working on this.

I am working to make a new release of towncrier, and the plan is to drop 3.7 support.

So this PR should be even simpler.

I am happy to drop the `incremental` dependency from towncrier.
I think that managing it via a simple string is good enough.
The version should only be modified by the release manager and the release manager should be familiar with the versioning scheme.

I think that using `Version('towncrier', 19, 9, 0, release_candidate=1)` instead of just `19.9.0.rc1` is overengineering

**Comment by SmileyChris** [comment]
@adiroiban cool! I'd say merge this first then work on dropping the 3.7 support, otherwise I can update this after you've changed the automated tests so this doesn't fail when I remove the import try/excepts.

(also a review of #482 would be nice, it would be a good addition before the new release ;) )

File: src/towncrier/_shell.py [hunk 2]
```diff
@@ -13,14 +13,25 @@
 
 from click_default_group import DefaultGroup
 
-from ._version import __version__
 from .build import _main as _build_cmd
 from .check import _main as _check_cmd
 from .create import _main as _create_cmd
 
 
+try:
```
Comment (adiroiban) [inline src/towncrier/_shell.py:23]: Maybe we can still keep the `_version.py` file but this time just just a `_version` string variable.

I this way, we don't have to repeat the `importlib` code each time we want to fetch the version

File: src/towncrier/_shell.py [hunk 3]
```diff
@@ -13,14 +13,25 @@
...
+    from importlib.metadata import PackageNotFoundError, version
+except ImportError:  # < Python 3.8
+    from importlib_metadata import PackageNotFoundError, version  # type: ignore
+
+
+try:
```
Comment (adiroiban) [inline src/towncrier/_shell.py]: this code can be included in the updated `_version.py` code

File: src/towncrier/test/test_packaging.py [hunk 4]
```diff
@@ -1,22 +1,16 @@
...
+import towncrier
 
 
 class TestPackaging(TestCase):
-    def test_version_warning(self):
+    def no_version_attr(self):
         """
-        Import __version__ from towncrier returns an Incremental version object
-        and raises a warning.
+        towncrier.__version__ was deprecated, now no longer exists.
```
Comment (adiroiban) [inline src/towncrier/test/test_packaging.py:12]: not sure if `towncrier.__version__` was considered public or private API.

For private API, raising AttributeError is ok.
If it's public API, we need to raise a separate error with an info message.

File: src/towncrier/test/test_project.py [hunk 5]
```diff
@@ -46,25 +39,61 @@ def test_tuple(self):
         version = get_version(temp, "mytestproja")
         self.assertEqual(version, "1.3.12")
 
-    @skipIf(metadata_version is None, "Needs importlib.metadata.")
     def test_incremental(self):
         """
-        An incremental Version __version__  is picked up.
+        An incremental-like Version __version__  is picked up.
```
Comment (adiroiban) [inline src/towncrier/test/test_project.py:44]: what does `incremental-like` version means ?

I guess that this means a `__version__` object/instance that has the same public API/interface as incremental.
That is, it doesn't have to be `incremental` it only needs to behave like it.

File: src/towncrier/test/test_project.py [hunk 6]
```diff
@@ -46,25 +39,61 @@ def test_tuple(self):
...
         """
-        pkg = "../src"
+        temp = self.mktemp()
+        os.makedirs(temp)
+        os.makedirs(os.path.join(temp, "mytestprojinc"))
+
+        with open(os.path.join(temp, "mytestprojinc", "__init__.py"), "w") as f:
+            f.write(
+                """
+class Version:
```
Comment (adiroiban) [inline src/towncrier/test/test_project.py:53]: maybe a docstring can be added to infom that this tries to "emulate" an incremntal class.

File: src/towncrier/test/test_project.py (hunk 1 above)
Comment (adiroiban) [reply src/towncrier/test/test_project.py:90]: I think that it's ok to keep Exception for now.

Later we can look at raising ClickException or something like this for better CLI integration... but in a separate PR.

**Review by adiroiban — CHANGES_REQUESTED** [review]
The changes look good.

Only minor comments.

I think  that it would be nice to be able to still do `from towncrier._version import version` or something similar, without having to worry about importlib each time.

Now, I don't know if it should raise an exception or return a dummy string, when the version was not found... I would go with a "dummy/placeholder" string.

**Comment by adiroiban** [comment]
@SmileyChris #482 it's on my todo list... but we can trigger a release anytime.

File: src/towncrier/test/test_project.py (hunk 5 above)
Comment (SmileyChris) [reply src/towncrier/test/test_project.py:44]: Correct. The code doesn't check the instance of the __version__, just that it quacks like an incremental version.

**Comment by hynek** [comment]
JFTR: if you'd like to entertain the idea of using `hatch-vcs` to deduplicate version information, I'm happy to help. I used to be skeptical to the idea, but now I've moved all my projects to it. If configured correctly, it also allows for continuous uploads to Test PyPI: https://test.pypi.org/project/structlog/#history

I guess the "you" here is mostly towards @adiroiban since he's doing all the release work and @glyph because he's The Twisted King.

**Comment by adiroiban** [comment]
To me `hatch-vcs` is over-engineering.
I am not sure what problem it tries to solve.

Is it that hard to keep the version as a text inside a .toml file? :)

**Comment by hynek** [comment]
Well, it's not. :) But it deduplicates the version information to one place (we do git tags too) and enables the continuous uploads I've mentioned. But obviously, if you don't want it, it's not happening. Just wanted to offer. :)

**Comment by hynek** [comment]
this needs a rebase / conflict resolution

**Comment by SmileyChris** [comment]
Removed the python 3.7 related fallbacks, updated the version in the toml, and rebased against trunk.
(and pushing again with no change because the `checknewsfragment` test failed for no good reason 😖 )

File: src/towncrier/__init__.py [hunk 7]
```diff
@@ -4,29 +4,3 @@
...
-
-    warnings.warn(
-        "Accessing towncrier.__version__ is deprecated and will be "
-        "removed in a future release. Use importlib.metadata directly "
-        "to query for towncrier's packaging metadata.",
-        DeprecationWarning,
-        stacklevel=2,
-    )
-
-    return __version__
```
Comment (hynek) [inline src/towncrier/__init__.py:32]: I'm not sure this should be removed. It seems like there's still a bit of software relying on `pkg.__version__` such that I've starte de-deprecating it in my own packages. See also https://github.com/python-attrs/attrs/issues/1136

(to be clear: it has to be ported, not just left around ;))

File: src/towncrier/_project.py [hunk 8]
```diff
@@ -38,47 +38,45 @@ def _get_package(package_dir: str, package: str) -> ModuleType:
 
 
 def get_version(package_dir: str, package: str) -> str:
-    module = _get_package(package_dir, package)
+    version: Any
```
Comment (hynek) [inline src/towncrier/_project.py:41]: Why Any? Any opts `version` completely out of type checking, is there a rescue for that?

File: src/towncrier/_project.py [hunk 9]
```diff
@@ -38,47 +38,45 @@ def _get_package(package_dir: str, package: str) -> ModuleType:
...
 
 def get_project_name(package_dir: str, package: str) -> str:
     module = _get_package(package_dir, package)
-
     version = getattr(module, "__version__", None)
+    # Incremental has support for package names, try duck-typing it.
+    try:
+        return str(version.package)  # type: ignore
+    except AttributeError:
+        pass
```
Comment (hynek) [inline src/towncrier/_project.py:80]:
```suggestion
    with contextlib.suppress(AttributeError):
        return str(version.package)  # type: ignore
```

is more idiomatic, I think

File: src/towncrier/_shell.py (hunk 2 above)
Comment (hynek) [reply src/towncrier/_shell.py:23]: We can do this lazily using `towncrier.__version__` if the block that I've commented on earlier is ported and kill two birds with one stone.

File: src/towncrier/test/test_packaging.py [hunk 10]
```diff
@@ -1,22 +1,16 @@
...
-from incremental import Version
 from twisted.trial.unittest import TestCase
 
-from towncrier._version import _hatchling_version
+import towncrier
 
 
 class TestPackaging(TestCase):
-    def test_version_warning(self):
+    def no_version_attr(self):
```
Comment (hynek) [inline src/towncrier/test/test_packaging.py:10]: I don't think this runs if it doesn't have a `test_` prefix.

**Review by hynek — CHANGES_REQUESTED** [review]
I've added a few notes, but ultimately I'm unconditionally yielding the decisions to @adiroiban since this has been his dream for years. ;)

File: src/towncrier/_project.py [hunk 11]
```diff
@@ -38,47 +38,45 @@ def _get_package(package_dir: str, package: str) -> ModuleType:
...
 def get_version(package_dir: str, package: str) -> str:
-    module = _get_package(package_dir, package)
+    version: Any
 
+    module = _get_package(package_dir, package)
     version = getattr(module, "__version__", None)
-
     if not version:
-        raise Exception("No __version__, I don't know how else to look")
+        try:
```
Comment (adiroiban) [inline src/towncrier/_project.py:46]: it looks like this is not only about removing incremental, but also about using `importlib_metadata` to get the version of  the target project

I think is best to do this as part of a separate PR ..l like https://github.com/twisted/towncrier/pull/502

**Comment by sigma67** [comment]
Hi @adiroiban @SmileyChris is this PR stale and ready to be picked up by someone else? Or are you planning to continue work?

File: src/towncrier/_project.py [hunk 12]
```diff
@@ -38,47 +38,45 @@ def _get_package(package_dir: str, package: str) -> ModuleType:
...
     if isinstance(version, tuple):
         return ".".join(map(str, version)).strip()
 
+    # Try duck-typing as an Incremental version.
+    if hasattr(version, "base"):
+        try:
+            version = str(version.base()).strip()
+            # Incremental uses `X.Y.rcN`.
+            # Standardize on importlib (and PEP440) use of `X.YrcN`:
+            return version.replace(".rc", "rc")  # type: ignore
```
Comment (twm) [inline src/towncrier/_project.py:63]: FYI this has been fixed on the Incremental side.

**Comment by adiroiban** [comment]
Hi @sigma67 

This PR is stale.

If you have time, feel free to continue working on it.

Thanks

File: src/towncrier/__init__.py (hunk 7 above)
Comment (sigma67) [reply src/towncrier/__init__.py:32]: re-added #627

File: src/towncrier/_project.py (hunk 8 above)
Comment (sigma67) [reply src/towncrier/_project.py:41]: changed to str #627

File: src/towncrier/_project.py (hunk 9 above)
Comment (sigma67) [reply src/towncrier/_project.py:80]: done  #627

File: src/towncrier/test/test_packaging.py (hunk 10 above)
Comment (sigma67) [reply src/towncrier/test/test_packaging.py:10]: changed test to check for deprecationwarning instead #627

File: src/towncrier/_project.py (hunk 11 above)
Comment (sigma67) [reply src/towncrier/_project.py:46]: Stale comment as linked PR has been merged. Updated and integrated latest trunk in  #627

File: src/towncrier/_project.py (hunk 12 above)
Comment (sigma67) [reply src/towncrier/_project.py:63]: Not sure what you are referring to. Should anything in this PR be changed as a result?

File: src/towncrier/_shell.py (hunk 2 above)
Comment (sigma67) [reply src/towncrier/_shell.py:23]: `_version.py` restored with contents

```
__version__ = "23.11.1.dev0"
```

**Comment by sigma67** [comment]
Hi @adiroiban, thanks, I've picked it up in #627

**Comment by adiroiban** [comment]
I am closing this as the current effort is with https://github.com/twisted/towncrier/pull/627
